confluent_kafka/schema_registry/schema_registry_client.py

```diff
@@ -0,0 +1,701 @@
+
+    def put(self, url, body=None):
+        return self.send_request(url, method='PUT', body=body)
+
+    def send_request(self, url, method, body=None, query=None):
+        """
+        Sends HTTP request to the SchemaRegistry.
+
+        All unsuccessful attempts will raise a SchemaRegistryError with the
+        response contents. In most cases this will be accompanied with a
+        Schema Registry supplied error code.
+
+        In the event the response is malformed an error_code of -1 will be used.
+
+        Args:
+            url (str): Request path
+            method (str): HTTP method
+            body (str): Request content
+            query (dict): Query params to attach to the URL
+
+        Returns:
+            dict: Schema Registry response content.
+        """
+        _headers = self.default_headers
+
+        if body is not None:
+            body = json.dumps(body)
+            _headers = {'Content-Length': str(len(body)),
+                        'Content-Type': "application/vnd.schemaregistry.v1+json"}
+
+        response = self.session.request(
+            method, url="/".join([self.base_url, url]),
+            headers=_headers, data=body, params=query)
+
+        try:
+            if 200 <= response.status_code <= 299:
+                return response.json()
+            raise SchemaRegistryError(response.status_code,
+                                      response.json().get('error_code'),
+                                      response.json().get('message'))
+        # Schema Registry may return HTML when it hits unexpected errors
+        except Exception:
+            raise SchemaRegistryError(response.status_code,
+                                      -1,
+                                      "Unknown Schema Registry Error: "
+                                      + response.content)
+
+
+class _SchemaCache(object):
+    """
+    Thread-safe Cache for use with the Schema Registry Client.
+
+    This cache may be used to retreive schema ids, schemas or to check
+    known subject membership.
+
+    """
+
+    def __init__(self):
+        self.lock = Lock()
+        self.schema_id_index = {}
+        self.schema_index = {}
+        self.subject_schemas = defaultdict(set)
+
+    def set(self, schema_id, schema, subject_name=None):
+        """
+        Add a Schema identified by schema_id to the cache.
+
+        Args:
+            schema_id (int): Schema's registration id
+            schema (Schema): Schema instance
+            subject_name(str): Optional, subject schema is registered under
+
+        Returns:
+            int: The schema_id
+        """
+        with self.lock:
+            # Don't overwrite existing keys
+            if schema_id not in self.schema_id_index:
+                self.schema_id_index[schema_id] = schema
+                if subject_name is not None:
+                    self.subject_schemas[subject_name].add(schema)
+            if schema not in self.schema_index:
```

Comment:
i think this if should be indented, otherwise it's possible to have inconsistent id -> schema and schema -> id.